### PR TITLE
[CuteDSL] Add benchmarking, CE memory optimizations, and match Triton num_warps

### PR DESCRIPTION
**Pull request by justinhh4**
## Summary

Follow-up to #1279 (the merged CuTe DSL cross-entropy PR), addressing @vaibhavjindal's review
follow-ups. The headline is a **memory optimization** to the CuTe DSL CE backward that brings its peak
memory to **exact Triton parity** (it was ~2× before). Also adds the **CuTe DSL benchmarking scripts**
(like the cuTile ones), and — in this description — **B200 speed + memory numbers** vs the fully
optimized Triton CE, an **H100 section** (memory parity confirmed; Triton slightly ahead on speed), and
an **explanation of the source of the gains**.

**TL;DR on speed:** against Triton with every CE optimization applied (through #1268) on the standard
last-layer step, the two backends are **at parity** — CuTe DSL edges bf16 by ~1.0–1.18× and is within
noise on fp32. This isn't a speed PR; it's a **memory-parity** PR (plus benchmark tooling).

Reviewer asks addressed:
1. ✅ **Memory** — measured; the backward is now Triton-parity (details + fix below, tables below).
2. ✅ **H100 speed/memory** — filled below from an H100 run on this branch (full sweep, same tables as
   B200). Memory: exact Triton parity at every point. Speed: after a Hopper warp-count fix (arch-aware
   num_warps, mirroring Triton), bf16 is parity-to-a-win on vocab / ~0.90× on the BT sweep and fp32 is
   parity — see the H100 section.
3. ✅ **Source of the gains** — explained below.
4. ✅ **CuTe DSL benchmark scripts** — added (`run_cutedsl_compare.py`, mirroring
   `run_cutile_compare.py`).

### The optimization: CE backward memory parity

As merged in #1279, the CuTe DSL CE backward, on the scalar-grad path (`reduction='mean'|'sum'`, i.e.
the normal training case), computed `_input * grad_output` — which **allocates a second `BT×V`
buffer**. The Triton CE instead scales the saved gradient **in place** over the logits via a raw
element-wise kernel, so it only ever holds **1× the logits tensor**. Net effect: the CuTe DSL CE used
**~2× the peak memory** of Triton in the fwd+bwd step (e.g. bf16, BT=8192, V=128256: **4008 → 2004 MB**).

**Fix** (`src/liger_kernel/ops/cutedsl/ops/cross_entropy.py`): scale the gradient in place, reusing the
same `element_mul_kernel` the Triton CE uses. A *raw* Triton kernel is used (rather than
`_input *= grad_output`) on purpose: an in-place torch mul on the tensor returned from `forward` bumps
its autograd version counter and trips backward-through-backward; the raw kernel writes through the
pointer without that bookkeeping — exactly how the Triton CE backward dodges the same issue. Peak
memory now **equals Triton across the entire sweep** (tables below).

## Benchmarks — CuTe DSL vs Triton (NVIDIA B200)

Speed: median of 40 iters, warmup 20 (CUDA events). `full` = fwd + `loss.backward()` 
Peak memory: median of 5, `torch.cuda.max_memory_allocated`. **The Triton baseline is the fully
optimized CE — every merged CE perf change plus #1268** (exp2 #1266, dtype-aware `num_warps` #1267, single post-loop `dx_y` correction #1268) — so this is CuTe DSL vs Triton at its best.
**speedup = Triton ÷ CuteDSL** (>1.00 → CuteDSL faster). Reference config: BT=8192, V=128256.

### bf16

**Speed — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (ms) | triton (ms) | speedup |
|------:|-------------:|------------:|--------:|
| 32000 | 0.4110 | 0.4519 | 1.100× |
| 50304 | 0.5673 | 0.6717 | 1.184× |
| 102400 | 0.9719 | 1.1194 | 1.152× |
| 128256 | 1.1660 | 1.1863 | 1.017× |
| 152064 | 1.3395 | 1.4501 | 1.083× |
| 201088 | 1.7105 | 1.9816 | 1.158× |
| 262144 | 2.2184 | 2.3193 | 1.045× |

**Speed — full — BT sweep (V=128256)**
| BT | cutedsl (ms) | triton (ms) | speedup |
|---:|-------------:|------------:|--------:|
| 1024 | 0.3454 | 0.3412 | 0.988× |
| 2048 | 0.4655 | 0.4660 | 1.001× |
| 4096 | 0.6969 | 0.7099 | 1.019× |
| 8192 | 1.1651 | 1.1884 | 1.020× |
| 16384 | 2.0957 | 2.1927 | 1.046× |
| 32768 | 4.0186 | 4.1108 | 1.023× |
| 65536 | 7.7521 | 7.9471 | 1.025× |

**Peak memory — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (MB) | triton (MB) |
|------:|-------------:|------------:|
| 32000 | 500 | 500 |
| 50304 | 786 | 786 |
| 102400 | 1600 | 1600 |
| 128256 | 2004 | 2004 |
| 152064 | 2376 | 2376 |
| 201088 | 3142 | 3142 |
| 262144 | 4096 | 4096 |

**Peak memory — full — BT sweep (V=128256)**
| BT | cutedsl (MB) | triton (MB) |
|---:|-------------:|------------:|
| 1024 | 250 | 250 |
| 2048 | 502 | 502 |
| 4096 | 1002 | 1002 |
| 8192 | 2004 | 2004 |
| 16384 | 4008 | 4008 |
| 32768 | 8017 | 8017 |
| 65536 | 16034 | 16033 |

### fp32

**Speed — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (ms) | triton (ms) | speedup |
|------:|-------------:|------------:|--------:|
| 32000 | 0.7377 | 0.6491 | 0.880× |
| 50304 | 0.9406 | 0.9516 | 1.012× |
| 102400 | 1.4447 | 1.6519 | 1.143× |
| 128256 | 1.8539 | 1.8128 | 0.978× |
| 152064 | 2.2744 | 2.2852 | 1.005× |
| 201088 | 3.1161 | 3.1875 | 1.023× |
| 262144 | 4.0071 | 3.9991 | 0.998× |

**Speed — full — BT sweep (V=128256)**
| BT | cutedsl (ms) | triton (ms) | speedup |
|---:|-------------:|------------:|--------:|
| 1024 | 0.4252 | 0.4224 | 0.993× |
| 2048 | 0.6288 | 0.6250 | 0.994× |
| 4096 | 1.0395 | 1.0222 | 0.983× |
| 8192 | 1.8552 | 1.8100 | 0.976× |
| 16384 | 3.5445 | 3.4421 | 0.971× |
| 32768 | 6.8232 | 6.5943 | 0.966× |
| 65536 | 13.3974 | 13.1099 | 0.979× |

**Peak memory — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (MB) | triton (MB) |
|------:|-------------:|------------:|
| 32000 | 1000 | 1000 |
| 50304 | 1572 | 1572 |
| 102400 | 3200 | 3200 |
| 128256 | 4008 | 4008 |
| 152064 | 4752 | 4752 |
| 201088 | 6284 | 6284 |
| 262144 | 8192 | 8192 |

**Peak memory — full — BT sweep (V=128256)**
| BT | cutedsl (MB) | triton (MB) |
|---:|-------------:|------------:|
| 1024 | 502 | 502 |
| 2048 | 1002 | 1002 |
| 4096 | 2004 | 2004 |
| 8192 | 4008 | 4008 |
| 16384 | 8016 | 8016 |
| 32768 | 16033 | 16033 |
| 65536 | 32066 | 32065 |

### Notes
- **bf16**: CuTe DSL edges out the fully optimized Triton on the vocab sweep (**1.02–1.18×**) and is at
  parity on the BT sweep (**0.99–1.05×**). These numbers line up with #1279's original table (which
  compared against the same optimized Triton), now on the memory-fixed kernel.
- **fp32**: essentially parity (**0.97–1.14×**); Triton is slightly ahead on the BT sweep. fp32 halves
  the 128-bit vectorization width and doubles the bytes streamed, so the kernel is fully DRAM-bound and
  both backends saturate the same bandwidth.
- **Memory**: CuTe DSL now **matches Triton exactly** at every point (1× the logits tensor, since the
  backward now scales the gradient in place instead of allocating a second `BT×V` buffer) — **this is
  the headline of the PR.**

### bf16 (H100)

**Speed — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (ms) | triton (ms) | speedup |
|------:|-------------:|------------:|--------:|
| 32000 | 0.7215 | 0.7975 | 1.105× |
| 50304 | 1.0494 | 1.1438 | 1.090× |
| 102400 | 1.9960 | 1.8839 | 0.944× |
| 128256 | 2.4283 | 2.2067 | 0.909× |
| 152064 | 2.8312 | 2.7233 | 0.962× |
| 201088 | 3.6500 | 3.6274 | 0.994× |
| 262144 | 4.6563 | 4.6008 | 0.988× |

**Speed — full — BT sweep (V=128256)**
| BT | cutedsl (ms) | triton (ms) | speedup |
|---:|-------------:|------------:|--------:|
| 1024 | 0.6004 | 0.5601 | 0.933× |
| 2048 | 0.8644 | 0.7973 | 0.922× |
| 4096 | 1.3858 | 1.2731 | 0.919× |
| 8192 | 2.4278 | 2.2088 | 0.910× |
| 16384 | 4.5233 | 4.0762 | 0.901× |
| 32768 | 8.7104 | 7.8105 | 0.897× |
| 65536 | 17.0571 | 15.2874 | 0.896× |

**Peak memory — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (MB) | triton (MB) |
|------:|-------------:|------------:|
| 32000 | 500 | 500 |
| 50304 | 786 | 786 |
| 102400 | 1600 | 1600 |
| 128256 | 2004 | 2004 |
| 152064 | 2376 | 2376 |
| 201088 | 3142 | 3142 |
| 262144 | 4096 | 4096 |

**Peak memory — full — BT sweep (V=128256)**
| BT | cutedsl (MB) | triton (MB) |
|---:|-------------:|------------:|
| 1024 | 250 | 250 |
| 2048 | 502 | 502 |
| 4096 | 1002 | 1002 |
| 8192 | 2004 | 2004 |
| 16384 | 4008 | 4008 |
| 32768 | 8017 | 8017 |
| 65536 | 16034 | 16033 |

### fp32 (H100)

**Speed — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (ms) | triton (ms) | speedup |
|------:|-------------:|------------:|--------:|
| 32000 | 1.3057 | 1.0447 | 0.800× |
| 50304 | 1.9576 | 1.6267 | 0.831× |
| 102400 | 3.7017 | 3.6407 | 0.984× |
| 128256 | 4.5613 | 4.5006 | 0.987× |
| 152064 | 5.3596 | 5.3013 | 0.989× |
| 201088 | 6.9944 | 6.9316 | 0.991× |
| 262144 | 9.0163 | 8.9290 | 0.990× |

**Speed — full — BT sweep (V=128256)**
| BT | cutedsl (ms) | triton (ms) | speedup |
|---:|-------------:|------------:|--------:|
| 1024 | 0.8693 | 0.8518 | 0.980× |
| 2048 | 1.3944 | 1.3716 | 0.984× |
| 4096 | 2.4547 | 2.4148 | 0.984× |
| 8192 | 4.5600 | 4.5009 | 0.987× |
| 16384 | 8.7760 | 8.6826 | 0.989× |
| 32768 | 17.1979 | 17.0310 | 0.990× |
| 65536 | 34.0251 | 33.6152 | 0.988× |

**Peak memory — full (fwd+bwd) — vocab sweep (BT=8192)**
| vocab | cutedsl (MB) | triton (MB) |
|------:|-------------:|------------:|
| 32000 | 1000 | 1000 |
| 50304 | 1572 | 1572 |
| 102400 | 3200 | 3200 |
| 128256 | 4008 | 4008 |
| 152064 | 4752 | 4752 |
| 201088 | 6284 | 6284 |
| 262144 | 8192 | 8192 |

**Peak memory — full — BT sweep (V=128256)**
| BT | cutedsl (MB) | triton (MB) |
|---:|-------------:|------------:|
| 1024 | 502 | 502 |
| 2048 | 1002 | 1002 |
| 4096 | 2004 | 2004 |
| 8192 | 4008 | 4008 |
| 16384 | 8016 | 8016 |
| 32768 | 16033 | 16033 |
| 65536 | 32066 | 32065 |

## Source of the gains vs Triton

First, the honest framing: with the Triton CE fully optimized (through #1268) and the standard
last-layer step (grad_output = 1.0, where the backward is a no-op), **the two backends are close to
parity** — CuTe DSL wins bf16 by a modest ~1.0–1.18× and is within noise on fp32. The value of this PR
is the **memory parity fix**; the speed is a wash-to-slight-win. Where CuTe DSL does pull ahead (bf16),
it comes from the forward kernel:

1. **Explicit `cp.async` multi-stage smem pipeline** — 128-bit vectorized, L1-bypassing loads with a
   power-of-2 ring buffer, prologue prefetch, and `cp_async_wait_group` staging. Global loads are
   prefetched ahead of the online-softmax compute, hiding HBM latency somewhat better than
   compiler-scheduled loads on Blackwell (sm_100). This is the main bf16 edge.
2. **Lower fixed host overhead** — the three target-validation D2H syncs are batched into one
   (`torch.stack((count, max, min)).tolist()`), the `CUstream` is cached, and DLPack handles for
   unused optional outputs are reused. That constant ≈25–40 µs helps the small-vocab corners most.

Both backends run the *same* base-2 online-softmax with hardware `ex2.approx` (Triton via #1266), so
the math and numerics are identical — there is no algorithmic advantage, only scheduling/overhead. In
fp32 the kernel is fully DRAM-bound and that overhead is amortized away, so Triton (which does the same
in-place-gradient bandwidth work) matches or slightly beats CuTe DSL.

## Benchmark scripts

- **`benchmark/scripts/run_cutedsl_compare.py`** — Triton-vs-CuTe-DSL compare driver, mirroring
  `run_cutile_compare.py`. Runs `benchmark_<kernel>.py` twice under `LIGER_KERNEL_IMPL` and tags each
  series (`liger_triton` / `liger_cutedsl` / `torch`) into
  `benchmark/data/all_benchmark_data_cutedsl.csv`. Supports `cross_entropy`.
- **`benchmark/README.md`** — document both compare drivers (CuTile + CuTe-DSL).
- **`benchmark/data/all_benchmark_data_cutedsl.csv`** — sample CE dataset generated by the new driver
  on B200.

## Correctness

 `test/transformers/test_cutedsl_cross_entropy.py`: **158 passed, 1 skipped** on both B200 and H100 (the not-last-layer / scalar-grad tests exercise exactly the fixed backward path; `test_num_warps_matches_dtype_convention` now asserts the arch-aware warp count on each GPU).

### CODE REVIEW

File: src/liger_kernel/ops/cutedsl/ops/cross_entropy.py [hunk 1]
```diff
@@ -827,9 +842,23 @@ def cross_entropy_backward(_input, grad_output):
...
-    return _input * grad_output
+    # reduction in {mean, sum}: scalar upstream grad. Scale the saved gradient IN PLACE so we
+    # never materialize a second BT×V buffer (Triton-parity peak memory: 1x logits, not 2x).
+    # A raw Triton element-wise kernel is used instead of `_input *= grad_output` because an
+    # in-place torch mul on the tensor returned from forward bumps its autograd version counter
+    # and trips backward-through-backward anomalies; the raw kernel writes through the pointer
+    # without that bookkeeping — exactly how the Triton CE backward dodges the same issue.
+    BT, V = _input.shape
+    BLOCK_SIZE = min(_MAX_FUSED_SIZE, triton.next_power_of_2(V))
+    element_mul_kernel[(BT,)](
```
Comment (vaibhavjindal) [inline src/liger_kernel/ops/cutedsl/ops/cross_entropy.py]: Let's not use triton kernel in cutedsl code. This is a bit confusing and could lead to issues when triton is not installed.

Let's revert to the original code, or maybe write an equivalent kernel in cutedsl.

Comment (justinhh4) [reply src/liger_kernel/ops/cutedsl/ops/cross_entropy.py]: @vaibhavjindal makes sense. I have implemented a cutedsl version and got roughly the same speed and memory numbers. Let me know if there is anything else to change!

File: src/liger_kernel/ops/cutedsl/ops/cross_entropy.py [hunk 2]
```diff
@@ -142,6 +145,53 @@ def _advance(idx, n: cutlass.Constexpr):
...
+        cute.make_ptr(mX.element_type, gX.iterator.toint(), cute.AddressSpace.gmem, assumed_align=16),
+        cute.make_layout((V,)),
+    )
+    VEC = const_expr(128 // gX.element_type.width)
+    gXv = cute.tiled_divide(gX, (VEC,))
+    num_vec = V // VEC
+    x_frag = cute.make_rmem_tensor((VEC,), gX.element_type)
+
+    for i in cutlass.range(0, cute.ceil_div(num_vec, 1024)):
+        vec_idx = tid + i * 1024
```
Comment (kolehma8) [inline src/liger_kernel/ops/cutedsl/ops/cross_entropy.py:169]: where is the 1024 coming? If that is to max out the 4kB page size, can we make this dependent on the numerical precision.

File: test/transformers/test_cutedsl_cross_entropy.py [hunk 3]
```diff
@@ -359,17 +359,27 @@ def test_ce_noncontiguous_input_matches_triton(dtype):
...
 )
-def test_num_warps_matches_dtype_convention(dtype, expected_warps):
+def test_num_warps_matches_dtype_convention(dtype):
+    from liger_kernel.ops.utils import is_hip
+    from liger_kernel.utils import infer_device_arch
+
+    if is_hip():
+        expected_warps = 16
+    else:
+        is_blackwell = infer_device_arch().startswith("blackwell")
```
Comment (kolehma8) [inline test/transformers/test_cutedsl_cross_entropy.py:380]: nit, should these be moved to a utility function?

Comment (justinhh4) [reply test/transformers/test_cutedsl_cross_entropy.py:380]: infer_device_arch() is already a utility function

File: src/liger_kernel/ops/cutedsl/ops/cross_entropy.py (hunk 2 above)
Comment (justinhh4) [reply src/liger_kernel/ops/cutedsl/ops/cross_entropy.py:169]: 1024 is threads per block, so vec_idx is obtained through thread index tid and loop index i

**Review by kolehma8 — APPROVED** [review]
